digitalocean/csi-digitalocean
# Spec v0.3.0

This implements the following:

* Updated the CSI spec to `v0.3.0`. This also required us to bump protobuf `v1.1.0`. The changes can be seen here: https://github.com/digitalocean/csi-digitalocean/commit/ae9fb5de1a74173da0dc75c18578ed6eff85fdde and https://github.com/digitalocean/csi-digitalocean/commit/96f2c15102b5d22ed62998ff1bf6d2cc4d5afc59
* Satisfies the new interfaces by adding the  new methods. `Snapshot` related methods are added but we return `codes.Unimplemented`. The implementation will be added later. The following methods are added: `CreateSnapshot`, `DeleteSnapshot`, `ListSnapshots`, `NodeGetInfo` (this replaces `NodeGetId`, which will be deprecated in future releases). The change can be seen here: https://github.com/digitalocean/csi-digitalocean/commit/a328cbf011ae9eecab1900669a682e6774ad01bc
* Exposed the maximum number of volumes that can be attached to a node in `NodeGetInfo`. The change can be seen here: https://github.com/digitalocean/csi-digitalocean/commit/7e2db60f5b3b2fd712c3bcd36694e4b1d9ea5aa7
* Improve tests and extend the fake DO API to support new use cases and the new tests in `csi-sanity` package. I had to make couple of hacks into the driver to make the tests happy. I'll plan to remove them once those are fixed in upstream. This change can be seen here: https://github.com/digitalocean/csi-digitalocean/commit/31fc8b1652baee2c7ad8ad3de4af761c12075a4d
* The `Probe()` method has a `ready` flag to signal the readiness of the plugin. Our plugin is very simple and therefore doesn't require much of initialization, but I made sure it's ready after everything is set up (such as getting the metadata is finished and plugin is running). The change can be seen here: https://github.com/digitalocean/csi-digitalocean/commit/e04e67d2c1dfe7c08a1610624456c359e2d81c67 
* Add `Topology` awareness (**new** in `v0.3.0`, see: https://github.com/container-storage-interface/spec/pull/188). This allows us to provision and manage volumes in a better way (Such as defining the rack/zone/region). DigitalOcean volumes are always `region` bound, so we make sure that all request has the correct region information. The change for this can be seen here: https://github.com/digitalocean/csi-digitalocean/commit/79bb621b831cc18e2cc515e5ec48292d3db5cd4c
* Updated sidecars to their newer version. The change can be seen here: https://github.com/digitalocean/csi-digitalocean/commit/656f470b8c401868712e256f10238b4adb984689 The following has changed:
  * updated `csi-attacher`sidecar from `v0.2.0` to `v0.3.0`
  * updated `csi-provisioner` sidecar from `v0.2.0` to `v0.3.0`
  * updated `driger-restrar` sidecar from `v0.2.0` to `v0.3.0`

Please review by reading each indivudal single commit independently

We're going to release `csi-digitalocean` in version `v0.2.0` after merging this PR, to indicate that new features were added and that the CSI spec was bumped. 

## How to test this change:

On a new cluster [follow the installation instructions](https://github.com/digitalocean/csi-digitalocean#installing-to-kubernetes) until **step 2**.  For step 2, make sure to install the manifest in this PR:

```
kubectl apply -f https://raw.githubusercontent.com/digitalocean/csi-digitalocean/656f470b8c401868712e256f10238b4adb984689/deploy/kubernetes/releases/csi-digitalocean-dev.yaml
```

## How to upgrade from csi-digitalocean `v0.1.x` to `v0.2.0` (the version based on this PR)

You can do this by updating the images. Run the following code to update all components:

Update plugin:

```
$ kubectl set image daemonset csi-doplugin csi-doplugin=digitalocean/do-csi-plugin:dev -n kube-system
$ kubectl set image statefulset csi-provisioner-doplugin digitalocean-csi-plugin=digitalocean/do-csi-plugin:dev -n kube-system && kubectl delete pods/csi-provisioner-doplugin-0 -n kube-system
$ kubectl set image statefulset csi-attacher-doplugin digitalocean-csi-plugin=digitalocean/do-csi-plugin:dev -n kube-system && kubectl delete pods/csi-attacher-doplugin-0  -n kube-system
```
Update sidecars

```
$ kubectl set image daemonset csi-doplugin driver-registrar=quay.io/k8scsi/driver-registrar:v0.3.0 -n kube-system
$ kubectl set image statefulset csi-provisioner-doplugin csi-provisioner=quay.io/k8scsi/csi-provisioner:v0.3.0 -n kube-system && kubectl delete pods/csi-provisioner-doplugin-0  -n kube-system
$ kubectl set image statefulset csi-attacher-doplugin csi-attacher=quay.io/k8scsi/csi-attacher:v0.3.0 -n kube-system && kubectl delete pods/csi-attacher-doplugin-0  -n kube-system
```
